lambda/translcore.ml

```diff
@@ -738,11 +812,13 @@ and transl_apply ~scopes
     | [] ->
         lapply lam (List.rev_map fst args)
   in
-  (build_apply lam [] (List.map (fun (l, x) ->
-                                   Option.map (transl_exp ~scopes) x,
-                                   Btype.is_optional l)
-                                sargs)
-     : Lambda.lambda)
+  let lam =
+    build_apply lam [] (List.map (fun (l, x) ->
+                            Option.map (transl_exp ~scopes) x,
+                            Btype.is_optional l)
+                          sargs)
+  in
+  maybe_region mode bound_modes lam
```

Comment:
You need a `maybe_region` whenever there is an expression in mode `Alloc_shared` that evaluates a subexpression in mode `Alloc_local`, because only then is there a possibility of a local allocation that must be deleted.

This can only happen on a handful of expressions. Let and apply are the most important cases, but there are a couple of others currently missing (Match, in particular)